apps/prairielearn/src/models/lti-user.ts

```diff
@@ -0,0 +1,108 @@
+import { z } from 'zod';
+
+import { callRow } from '@prairielearn/postgres';
+
+import { insertAuditEvent } from './audit-event.js';
+import { ensureEnrollment } from './enrollment.js';
+import { selectInstitutionByShortName } from './institution.js';
+import { insertUserLti, selectUserByUidAndInstitution, updateUserName } from './user.js';
+
+export interface LtiUserResult {
+  userId: string;
+  hasAccess: boolean;
+}
+
+export async function selectOrInsertAndEnrollLtiUser({
+  uid,
+  name,
+  lti_course_instance_id,
+  lti_user_id,
+  lti_context_id,
+  req_date,
+}: {
+  uid: string;
+  name: string | null;
+  lti_course_instance_id: string;
+  lti_user_id: string;
+  lti_context_id: string;
+  req_date: Date;
+}): Promise<LtiUserResult> {
+  const { id: institution_id } = await selectInstitutionByShortName({ short_name: 'LTI' });
+
+  let user = await selectUserByUidAndInstitution({
+    uid,
+    institution_id,
+  });
+
+  if (!user) {
+    user = await insertUserLti({
+      uid,
+      name,
+      lti_course_instance_id,
+      lti_user_id,
+      lti_context_id,
+      institution_id,
+    });
+
+    await insertAuditEvent({
+      action: 'insert',
+      table_name: 'users',
+      row_id: user.user_id,
+      new_row: user,
+      institution_id,
+      course_instance_id: lti_course_instance_id,
```

Comment:
I'm inclined to say that we should have a `course_id` if a `course_instance_id` is provided.

In PrairieTest, the insertion code will automatically walk IDs back up to fill in ones that are missing: https://github.com/PrairieLearnInc/PrairieTest/blob/58df32830bc4e46b3c1cc146aa64bcee09664947/src/models/audit-events.sql

It could be nice to do the same here.

Reply:
That's a good idea, I'll see if I can walk those up.

Reply:
Waiting on #12776